loopin.go

```diff
@@ -835,44 +831,117 @@ func (s *loopInSwap) waitForSwapComplete(ctx context.Context,
 		return err
 	}
 
-	htlcSpend := false
+	// We synchronize access to the sweepFee once the timeout transaction
+	// has been published. If we already published the timeout transaction
+	// we don't have to wait for the sweepFee to be set.
+	sweepFeeWg := sync.WaitGroup{}
+	sweepFeeWg.Add(1)
+	if sweepFee > 0 {
+		// We published the timeout transaction.
+		sweepFeeWg.Done()
+	}
+
+	// We monitor the htlc spend by the server or the timeout sweep in a
+	// separate go routine since we don't want to block on them in order to
+	// continue with successive loop ins once the invoice is settled.
+	executorWg.Add(1)
+	go func(s *loopInSwap) {
+		defer executorWg.Done()
+		spent := false
+		notifier := s.lnd.ChainNotifier
+		for !spent {
+			spendChan, spendErr, err := notifier.RegisterSpendNtfn(
+				ctx, htlcOutpoint, s.htlc.PkScript,
+				s.InitiationHeight,
+			)
+			if err != nil {
+				s.log.Infof("register spend ntfn: %v", err)
+				return
+			}
+
+			select {
+			// The htlc spend is confirmed. Inspect the spending tx
+			// to determine the final swap state.
+			case spendDetails := <-spendChan:
+				s.log.Infof("htlc spend by tx: %v",
+					spendDetails.SpenderTxHash)
+
+				// Inspect the input witness of the htlc spend
+				// and determine whether we swept with a success
+				// or a timeout transaction.
+				index := spendDetails.SpenderInputIndex
+				htlcInput := spendDetails.SpendingTx.TxIn[index]
+
+				// If the htlc was spent by the timeout
+				// transaction we wait until the sweep fee was
+				// set by the htlc publisher.
+				isSuccessWitness := s.htlc.IsSuccessWitness(
+					htlcInput.Witness,
+				)
+				if !isSuccessWitness {
+					sweepFeeWg.Wait()
+				}
+
+				err := s.processHtlcSpend(
+					ctx, isSuccessWitness, htlcValue,
+					sweepFee,
+				)
+				if err != nil {
+					s.log.Warnf("process spend error: %v",
+						err)
+					return
+				}
+
+				err = s.persistAndAnnounceState(ctx)
+				if err != nil {
+					s.log.Warnf("announce spend error: %v",
+						err)
+					return
+				}
+
+				spent = true
+
+			case spendError := <-spendErr:
+				s.log.Warnf("spend error: %v", spendError)
+				return
+
+			case <-ctx.Done():
+				s.log.Warnf("context canceled: %v", ctx.Err())
+				return
+			}
+		}
+	}(s)
+
 	invoiceFinalized := false
 	htlcKeyRevealed := false
-	for !htlcSpend || !invoiceFinalized {
+	for !invoiceFinalized {
 		select {
-		// Spend notification error.
-		case err := <-spendErr:
-			return err
-
 		// Receive block epochs and start publishing the timeout tx
 		// whenever possible.
 		case notification := <-s.blockEpochChan:
 			s.height = notification.(int32)
 
-			sweepFee, err = publishTxOnTimeout()
-			if err != nil {
-				return err
+			// If we haven't published the timeout transaction
+			// earlier we check if we can publish it. If we publish
+			// it here we mark the sweep fee as set so the htlc
+			// spend-processing goroutine can continue.
+			if sweepFee == 0 {
```

Comment:
Note that this changes the flow quite a bit. Previously we'd publush the timeout on each block. This is useful too, since the tx may get evicted from the mempool if there's a surge of incoming transactions on the network.

Reply:
That's a good catch, we wouldn't publish again after the first time. Will change that.